nexus/deployment/src/planner.rs

```diff
@@ -168,6 +162,126 @@ impl<'a> Planner<'a> {
+            &sleds_ineligible_for_services,
+        )?;
+
+        Ok(())
+    }
+
+    fn ensure_correct_number_of_nexus_zones(
+        &mut self,
+        sleds_ineligible_for_services: &BTreeSet<Uuid>,
+    ) -> Result<(), Error> {
+        // Bin every sled by the number of Nexus zones it currently has while
+        // counting the total number of Nexus zones.
+        let mut num_total_nexus = 0;
+        let mut sleds_by_num_nexus: BTreeMap<usize, Vec<Uuid>> =
+            BTreeMap::new();
+        for &sled_id in self.policy.sleds.keys() {
+            let num_nexus = self.blueprint.sled_num_nexus_zones(sled_id)?;
+            num_total_nexus += num_nexus;
+
+            // Only bin this sled if we're allowed to use it. If we have a sled
+            // we're not allowed to use that's already running a Nexus (seems
+            // fishy!), we counted its Nexus above but will ignore it here.
+            if !sleds_ineligible_for_services.contains(&sled_id) {
+                sleds_by_num_nexus.entry(num_nexus).or_default().push(sled_id);
+            }
+        }
+
+        // TODO-correctness What should we do if we have _too many_ Nexus
+        // instances? For now, just log it the number of zones any time we have
+        // at least the minimum number.
+        let nexus_to_add =
+            self.policy.target_nexus_zone_count.saturating_sub(num_total_nexus);
+        if nexus_to_add == 0 {
+            info!(
+                self.log, "sufficient Nexus zones exist in plan";
+                "desired_count" => self.policy.target_nexus_zone_count,
+                "current_count" => num_total_nexus,
+            );
+            return Ok(());
+        }
+
+        // Ensure we have at least one sled on which we can add Nexus zones. If
+        // we don't, we have nothing else to do. This isn't a hard error,
+        // because we might be waiting for NTP on all eligible sleds (although
+        // it would be weird, since we're presumably running from within Nexus
+        // on some sled).
+        if sleds_by_num_nexus.is_empty() {
+            warn!(self.log, "want to add Nexus zones, but no eligible sleds");
+            return Ok(());
+        }
+
+        // Build a map of sled -> new nexus zone count.
+        let mut sleds_to_change: BTreeMap<Uuid, usize> = BTreeMap::new();
+
+        'outer: for _ in 0..nexus_to_add {
+            // `sleds_by_num_nexus` is sorted by key already, and we want to
+            // pick from the lowest-numbered bin. We can just loop over its
+            // keys, expecting to stop on the first iteration, with the only
+            // exception being when we've removed all the sleds from a bin.
+            for (&num_nexus, sleds) in sleds_by_num_nexus.iter_mut() {
+                // TODO choose more smartly than effectively ordering by
+                // sled_id? See
+                // https://github.com/oxidecomputer/omicron/pull/4959#discussion_r1476795735.
```

Comment:
I'd probably strike this

Reply:
Because we don't want to do it, or because as we add more services and smarts to the planner it will be obvious that we need to do it? (E.g., if we duplicated this function for "cockroach" and "clickhouse", we'd end up ordering the sleds within each bin the same way.)

Reply:
Most of the time I feel like it's more distracting than useful to mention alternatives / possible future work in the source, unless it's really useful for understanding the current implementation.  If we want to keep track of this, I think we're better off filing an issue.  This is obviously subjective and it's just a suggestion.

Reply:
I reworded this to just briefly describe what it's currently doing